keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -0,0 +1,241 @@
+* [Replace top-level fields with per version fields](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresource-conversion-webhook.md#top-level-fields-to-per-version-fields)
+* Restrict OpenAPI per [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
+
+## Test Plan
+
+### Integration tests for GA
+
+TODO: complete this list
+
+* [Pruning for CustomResources KEP](https://github.com/kubernetes/enhancements/pull/709)
+  * TODO
+* [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
+  * TODO
+* CRD versioning/conversion (https://github.com/kubernetes/kubernetes/issues/64136):
+  * Ensure what is persisted in etcd matches the storage version
+  * Set up a CRD, persist some data, changed the version, and ensure the previously persisted data is readable
+  * Ensure discovery docs track a CRD through creation, version addition, version removal, and deletion
+  * Ensure `spec.served` is respected
+
+### e2e tests for GA
+
+TODO: complete this list
+
+* Ensure v1beta1 and v1 CRDs are round trippable
+* Ensure CR watch observes events on create/update/delete
+* Ensure a CR watch is correctly terminated if CRD is deleted
+
+### Conformance plan for GA
+
+TODO: complete this list
+
+* Promote existing e2e tests:
+  * [CustomResourceDefinition Watch](https://github.com/kubernetes/kubernetes/blob/36ccff1b27b6aee5c5599334842cb2b8cf8063f3/test/e2e/apimachinery/crd_watch.go#L37)
+  * [CustomResourcePublishOpenAPI](https://github.com/kubernetes/kubernetes/blob/36ccff1b27b6aee5c5599334842cb2b8cf8063f3/test/e2e/apimachinery/crd_publish_openapi.go#L52)
+  * [CustomResourceConversionWebhook](https://github.com/kubernetes/kubernetes/blob/36ccff1b27b6aee5c5599334842cb2b8cf8063f3/test/e2e/apimachinery/crd_conversion_webhook.go#L81)
+
+### Scale Targets for GA
+
+* TODO quantify: Read/write latency of CRDs within X% of native Kubernetes types
+* TODO quantify: Latency degrades less than X% for up to 100k Custom Resources per CRD kind
+* TODO quantify: Webhook conversion QPS of a noop converter is within X% of QPS with no webhook
+* Coordinate with sig-scalability
+
+## Graduation Criteria
+
+To mark these as complete, all of the above features need to be implemented. An
+umbrella issue is tracking all of these changes. Also there need to be
+sufficient tests for any of these new features and all existing features and
+documentation should be completed for all features.
+
+See [umbrella issue](https://github.com/kubernetes/kubernetes/issues/58682) for status.
+
+## Post-GA tasks
+
+See the [umbrella issue](https://github.com/kubernetes/kubernetes/issues/58682) for details on Post-GA tasks.
+
+### Arbitrary subresources as Post-GA task
```

Comment:
as nice as this is as a sketch, I would remove it from this KEP. This topic will move into its own KEP as soon as somebody seriously works on it. We have decided to do this post-GA though. So no need here yet.

Reply:
Sounds good. We wanted to think through the basics to make sure CRD API v1 would be compatible with adding arbitrary subresources later. But I agree that it doesn't need to be here in the KEP. We can discuss the details with those interested separately.

Reply:
We could start a gdoc at some point to work towards a 1.16 KEP for that.